k8s: Scenario guides (#72095)
ansible/ansible: docs/docsite/rst/scenario_guides/kubernetes_scenarios/scenario_k8s_object.rst

.. _k8s_object_template:

*******************
Creating K8S object
*******************

.. contents:: Topics

Introduction
============

This guide will show you how to utilize Ansible to create Kuberentes objects such as Pods, Deployments, and Secrets.

Scenario Requirements
=====================

* Software

    * Ansible 2.9.10 or later must be installed

>     * The Python modules ``openshift`` and ``kubernetes`` must be installed on the Ansible (or Target host if not executing against localhost)

Reviewer: `kubectl` for connection plugin again?

    * Kubernetes Cluster

* Access / Credentials

    * Kubeconfig configured with the given Kubernetes cluster

Assumptions
===========

- User has required level of authorization to create, delete and update resources on the given Kuberentes cluster.

Caveats
=======

> - NA

Reviewer: Mention that this content is being migrated to kubernetes.core.

Example Description
===================

In this use case / example, we will be creatin a Pod in the given Kubernetes Cluster.  The following Ansible playbook showcases the basic parameters that are needed for this.

.. code-block:: yaml

    ---
    - hosts: localhost
      collections:
        - community.kubernetes
      tasks:
      - name: Create a pod
        k8s:
          state: present
          definition:
            apiVersion: v1
            kind: Pod
            metadata:
              name: "utilitypod-1"
              namespace: default
              labels:
                app: galaxy
            spec:
              containers:
              - name: utilitypod
                image: busybox

Since Ansible utilizes the Kubernetes API to perform actions, in this use case we will be connecting directly to the Kubernetes cluster.

> To begin, there are a few bits of information we will need. Here you are using Kubeconfig which is pre-configured in your machine. The Kubeconfig is generally located at ``~/.kube/config``.It highly recommended to store sensetive information such as password, user certificates in a more secure fashion using :ref:`ansible-vault` or using `Ansible Tower credentials <https://docs.ansible.com/ansible-tower/latest/html/userguide/credentials.html>`_.

Reviewer: ```suggestion
To begin, there are a few bits of information we will need. Here you are using Kubeconfig which is pre-configured in your machine. The Kubeconfig is generally located at ``~/.kube/config``. It highly recommended to store sensitive information such as password, user certificates in a more secure fashion using :ref:`ansible-vault` or using `Ansible Tower credentials <https://docs.ansible.com/ansible-tower/latest/html/userguide/credentials.html>`_.
```
Missing a space and a typo.

Now you need to supply the information about the Pod which will be created. Using ``definition`` parameter of ``community.kubernetes.k8s`` module, you are specifying `PodTemplate <https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates>`_. This PodTemplate is identical to what you provide to ``kubectl`` command.

What to expect
--------------

- You will see a bit of JSON output after this playbook completes. This output shows various parameters that are returned from the module and from cluster about the newly created Pod.

> .. code-block:: yaml

Reviewer: The block is marked for YAML syntax but looks like JSON to me.

                "serviceAccount": "default",
                "serviceAccountName": "default",
                "terminationGracePeriodSeconds": 30,
                "tolerations": [
                    {
                        "effect": "NoExecute",
                        "key": "node.kubernetes.io/not-ready",
                        "operator": "Exists",
                        "tolerationSeconds": 300
                    },
                    {
                        "effect": "NoExecute",
                        "key": "node.kubernetes.io/unreachable",
                        "operator": "Exists",
                        "tolerationSeconds": 300
                    }
                ],
                "volumes": [
                    {
                        "name": "default-token-6j842",
                        "secret": {
                            "defaultMode": 420,
                            "secretName": "default-token-6j842"
                        }
                    }
                ]
            },
            "status": {
                "phase": "Pending",
                "qosClass": "BestEffort"
            }
        }
    }

> - State is changed to ``True`` which notifies that the Pod creation started on the given cluster. This can take some time depending on your environment.

Reviewer: Lost me a bit. I don't see "State" in the above example.

Reviewer: So you mean the task example above this example? That is set to "present" not "True" though.

Author: Re-worded a little bit.

> - If you utilize the ``wait`` parameter, module will wait until Pod is created and in running state.

Reviewer: This seems lost down here. Maybe move this up above "What to Expect?"

Author: OK. I removed this.

> - If you utilize the ``wait`` parameter, module will wait until Pod is created and in running state.

Reviewer: ```suggestion
- If you utilize the ``wait`` parameter, the `k8s` module will wait until Pod is created and in running state.
```

Author: Removed.

.. _k8s_object_template:

*******************
Creating K8S object
*******************

> .. contents:: Topics

Reviewer 2: ```suggestion
.. contents::
   :local:
```

Introduction
============

> This guide will show you how to utilize Ansible to create Kuberentes objects such as Pods, Deployments, and Secrets.

Reviewer 2: ```suggestion
This guide will show you how to utilize Ansible to create Kubernetes objects such as Pods, Deployments, and Secrets.
```

Scenario Requirements
=====================

* Software

    * Ansible 2.9.10 or later must be installed

    * The Python modules ``openshift`` and ``kubernetes`` must be installed on the Ansible controller (or Target host if not executing against localhost)

    * Kubernetes Cluster

    * Kubectl binary installed on the Ansible controller

* Access / Credentials

    * Kubeconfig configured with the given Kubernetes cluster

Assumptions
===========

> - User has required level of authorization to create, delete and update resources on the given Kuberentes cluster.

Reviewer 2: ```suggestion
- User has required level of authorization to create, delete and update resources on the given Kubernetes cluster.
```

Caveats
=======

- community.kubernetes 1.1.0 is going to migrate to `kubernetes.core <https://github.com/ansible-collections/kubernetes.core>`_

Example Description
===================

> In this use case / example, we will be creatin a Pod in the given Kubernetes Cluster.  The following Ansible playbook showcases the basic parameters that are needed for this.

Reviewer 2: ```suggestion
In this use case / example, we will  create a Pod in the given Kubernetes Cluster.  The following Ansible playbook showcases the basic parameters that are needed for this.
```

.. literalinclude:: task_outputs/Add_a_pod_using_k8s.yaml

Since Ansible utilizes the Kubernetes API to perform actions, in this use case we will be connecting directly to the Kubernetes cluster.

> To begin, there are a few bits of information we will need. Here you are using Kubeconfig which is pre-configured in your machine. The Kubeconfig is generally located at ``~/.kube/config``. It highly recommended to store sensitive information such as password, user certificates in a more secure fashion using :ref:`ansible-vault` or using `Ansible Tower credentials <https://docs.ansible.com/ansible-tower/latest/html/userguide/credentials.html>`_.

Reviewer 2: ```suggestion
To begin, there are a few bits of information we will need. Here you are using Kubeconfig which is pre-configured in your machine. The Kubeconfig is generally located at ``~/.kube/config``. It  is highly recommended to store sensitive information such as password, user certificates in a more secure fashion using :ref:`ansible-vault` or using `Ansible Tower credentials <https://docs.ansible.com/ansible-tower/latest/html/userguide/credentials.html>`_.
```

> Now you need to supply the information about the Pod which will be created. Using ``definition`` parameter of ``community.kubernetes.k8s`` module, you are specifying `PodTemplate <https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates>`_. This PodTemplate is identical to what you provide to ``kubectl`` command.

Reviewer 2: ```suggestion
Now you need to supply the information about the Pod which will be created. Using ``definition`` parameter of  the ``community.kubernetes.k8s`` module, you specify `PodTemplate <https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates>`_. This PodTemplate is identical to what you provide to the  ``kubectl`` command.
```

What to expect
--------------

- You will see a bit of JSON output after this playbook completes. This output shows various parameters that are returned from the module and from cluster about the newly created Pod.

.. literalinclude:: task_outputs/Add_a_pod_using_k8s.json

> - In above example, 'changed' is ``True`` which notifies that the Pod creation started on the given cluster. This can take some time depending on your environment.

Reviewer 2: ```suggestion
- In the above example, 'changed' is ``True`` which notifies that the Pod creation started on the given cluster. This can take some time depending on your environment.
```

Troubleshooting
---------------

Things to inspect

- Check if the values provided for username and password are correct
- Check if the Kubeconfig is populated with correct values

.. seealso::

    `Kubernetes Python client <https://github.com/kubernetes-client/python>`_
        The GitHub Page of Kubernetes Python client
    `Kubernetes Python client - Issue Tracker <https://github.com/kubernetes-client/python/issues>`_

>         The issue tracker for theKubernetes Python client

Reviewer 2: ```suggestion
        The issue tracker for the Kubernetes Python client
```

    `Openshift Python client <https://github.com/openshift/openshift-restclient-python>`_

>         The GitHub Page of Openshift Dynamic API client

Reviewer 2: ```suggestion
        The GitHub Page of OpenShift Dynamic API client
```

    `Openshift Python client - Issue Tracker <https://github.com/openshift/openshift-restclient-python/issues>`_

>         The issue tracker for Openshift Dynamic API client

Reviewer 2: ```suggestion
        The issue tracker for OpenShift Dynamic API client
```
